README.md

<div style="text-align:center">
    <h1> cajache </h1>
    <img height="256px" src="https://i.gyazo.com/468a720ce6a70550f5430794e42da631.png" />
</div>


[![cajache package size](https://packagephobia.now.sh/badge?p=cajache)](https://packagephobia.now.sh/result?p=cajache)
[![cajache package size minzipped](https://badgen.net/bundlephobia/minzip/cajache)](https://badgen.net/bundlephobia/minzip/cajache)



**cajache** is a minimalistic javascript caching library.

- ⚡ Optimizes your projects reducing the number of HTTP request or heavy actions performed.
- 🚀 Lightweight.
- ⚪️ Zero dependencies.



<br>



<!-- TOC ignore:true -->
# Table of contents


<!-- TOC -->

- [Table of contents](#table-of-contents)
- [Import](#import)
- [Quick start](#quick-start)
- [Creating a new instance](#creating-a-new-instance)
- [Instance options](#instance-options)
- [Use cases](#use-cases)
  - [Cache HTTP requests](#cache-http-requests)
  - [Cache paginated HTTP requests](#cache-paginated-http-requests)
- [API](#api)
  - [.use](#use)
  - [.get](#get)
  - [.set](#set)
  - [.delete](#delete)
  - [.deleteAll](#deleteall)
  - [.setConfig](#setconfig)
  - [TTL watcher](#ttl-watcher)
- [Go to top](#go-to-top)

<!-- /TOC -->



<br>



# Import

```js
const cajache = require("cajache");
```



# Quick start

Permanent cache

```js
const myFetchFnc = axios.get("https://your.api/resource");

const response = await cajache.use(
    "cache_id_01",
    myFetchFnc
);
// The first time it will execute myFetchFnc.

const cachedResponse = await cajache.use(
    "cache_id_01",
    myFetchFnc
);
// The second time it will return cached value instead re-executing myFetchFnc.
```
<br/>

Temporal cache

```js
const myFetchFnc = axios.get("https://your.api/resource");

const response = await cajache.use(
    "cache_id_01",
    myFetchFnc,
    {
        ttl: 1000, // 1 second
    }
);
// The first time it will execute myFetchFnc.

// Sleep 3
await new Promise(res => setTimeout(res, 3000));

const nonCachedResponse = await cajache.use(
    "cache_id_01",
    myFetchFnc,
    {
        ttl: 1000, // 1 second
    }
);
// The second time it will NOT return cached value because it's expired.
```

<br/>

# Creating a new instance

```js
const cajache = require("cajache");
const myInstance = cajache.new();
```

<br/>

# Instance options

```js
const cajache = cajache(options);
```

| Instance option     | Type           		| Description
| :-----------:       |:-------------:		| :-----	
| ttl      		      | number              | Default `0`. Default TTL in miliseconds for all cached values. 0 = permanent
| checkInterval       | function      		| Default `1000 * 60` (1 min). Interval in miliseconds to check if cached values with `ttl` are expired.
| path	              | string	            | Dot path to the property that will be cached. Example: `"axiosResponse.data"`.
| condition	          | function            | On cache miss, this function will receive as argument the response of `fnc`. If it returns `true` the response will be cached, if it returns `false` it won't be cached.


<br>


# Use cases

## Cache HTTP requests

```js

// fetch 1: 248.659ms
// fetch 2 (cached): 0.015ms
// fetch 3 (cached): 0.008ms


console.time("fetch 1");

let characters = await cajache.use(
    "characters",
    () => axios.get("https://rickandmortyapi.com/api/character/14"),
);

console.timeEnd("fetch 1");
console.time("fetch 2 (cached)");

characters = await cajache.use(
    "characters",
    () => axios.get("https://rickandmortyapi.com/api/character/14"),
);

console.timeEnd("fetch 2 (cached)");



console.time("fetch 3 (cached)");
await cajache.use(
    "characters",
    () => axios.get("https://rickandmortyapi.com/api/character/14"),
);
console.timeEnd("fetch 3 (cached)");


```

## Cache paginated HTTP requests

```js

// fetch page 1: 284.629ms
// fetch page 2: 208.210ms
// fetch page 1 (cached): 0.018ms
// fetch page 2 (cached): 0.008ms


console.time("fetch page 1");

let characters_page1 = await cajache.use(
    ["characters", "page_1"],
    () => axios.get("https://rickandmortyapi.com/api/character/?page=1"),
);

console.timeEnd("fetch page 1");
console.time("fetch page 2");

let characters_page2 = await cajache.use(
    ["characters", "page_2"],
    () => axios.get("https://rickandmortyapi.com/api/character/?page=2"),
);

console.timeEnd("fetch page 2");



console.time("fetch page 1 (cached)");

characters_page1 = await cajache.use(
    ["characters", "page_1"],
    () => axios.get("https://rickandmortyapi.com/api/character/?page=1"),
);

console.timeEnd("fetch page 1 (cached)");
console.time("fetch page 2 (cached)");

characters_page2 = await cajache.use(
    ["characters", "page_2"],
    () => axios.get("https://rickandmortyapi.com/api/character/?page=2"),
);

console.timeEnd("fetch page 2 (cached)");

```



<br>



# API

## .use

Syntax:
```js
const cachedResponse: Promise = cajache.use(
    id: "myId",
    fetchFnc: () => stuff,
    options: {},
);
```

| Parameter     | Type           			| Description 	|
| :-----------: |:-------------:			| :-----		|
| id      		| string \| Array\<string\>	| Unique identifier of the cache entry.	If you pass an array like `["parent", "child"]` it will be treated as deep nested id. If parent is deleted all children will be deleted too. 
| fnc      		| function      			| Your function that will be cached. Can be async.
| options 		| object      				| [Same as instance options](#instance-options) (without `checkInterval`). If set, it will override instance options, otherwise it will use them.

<br />

**Example:**
```js
// Simple id
const response = await cajache.use(
    "characters",
    () => axios.get("https://you.api.com/characters"),
);
```

```js
// Nested id
const response = await cajache.use(
    ["location_2",  "characters", "page_3"],
    () => axios.get("https://you.api.com/location2/characters?page=3"),
);
```


<br />

**Example with expire:**

```js
const response = await cajache.use(
    ["location_2", "characters", "page_3"],
    () => axios.get("https://you.api.com/location2/characters?page=3"),
    {
        expire: 1000 * 30, // 30 seconds
    }
);
```

<br />

**Example with path:**

```js
const response = await cajache.use(
    ["location_2", "characters", "page_3"],
    () => axios.get("https://you.api.com/location2/characters?page=3"),
    {
        path: "character.name",
    }
);
```

<br />

**Example with condition:**

```js
const response = await cajache.use(
    ["location_2", "characters", "page_3"],
    () => axios.get("https://you.api.com/location2/characters?page=3"),
    {
        condition: apiRes => apiRes.isError === false,
    }
);
```



<br/>



## .get

Syntax:
```js
cajache.get(id);
```


| Parameter     | Type           			| Description 	|
| :-----------: |:-------------:			| :-----		|
| id      		| string \| Array\<string\>	| Unique identifier of the cache entry

<br/>
Examples:
```js
const characters = cajache.get("characters");
```

Or...

```js
const location2_characters_page3 = cajache.get(["location_2", "characters", "page_3"]);
```



<br/>



## .set

Syntax:

```js
cajache.set(id, value );
```


| Parameter     | Type           			| Description 	|
| :-----------: |:-------------:			| :-----		|
| id      		| string \| Array\<string\>	| Unique identifier of the cache entry
| value 		| any	      				| Value you want to set

<br/>

Examples:
```js
cajache.set("characters", {...} );
```

Or...

```js
cajache.set(["location_2", "characters", "page_3"], {...} );
```


<br/>



## .delete

Syntax;
```js
cajache.delete(id);
```

| Parameter     | Type           			| Description 	|
| :-----------: |:-------------:			| :-----		|
| id      		| string \| Array\<string\>	| Unique identifier of the cache entry you want to delete.

<br/>

Delete `location_2` cache entry (and his childrens):
```js
cajache.delete("location_2");
```

Delete `location_2.characters.page_3` cache entry (and his childrens):





<br/>



## .deleteAll


Deletes all cache entries of the instance.
```js
cajache.deleteAll();
```

<br/>



## .setConfig


Sets the instance config.

```js
cajache.setConfig(key, value);
```

Keys and values are the [same as instance options](#instance-options).
<br/>

Example: 
```js
cajache.setConfig("checkInterval", 1000 * 60 * 5); // 5 minutes
```



---



## TTL watcher

- It starts automatically when you create any instance.
- It will only iterate over all cache entries with TTL.
- It will delete all expired cache entries every `config.checkInterval` milliseconds.
- It will check the instance config (`config?.checkInterval`) after each iteration to stop or continue the loop.
- If the `checkInterval` is changed it will not not be effective until the next iteration.



---



# <a name='table-of-contents'></a>[Go to top](#table-of-contents) 




CHANGELOG.md
# Changelog



## [2.0.2] - 2022-12-13

### Fixed
- `.delete` method was not deleting the cache entry.

### Added
- Tests for `.delete` and `.deleteAll` method.



## [2.0.1] - 2022-12-13

### Fixed
- `ttl` item on TTL watcher was being duplicated.
- Removed unnecesary console.log messages.



## [2.0.0] - 2022-12-13

### Breaking changes from  1.x.x to 2.0.0

- The `expire` option has been renamed to `ttl`. Instead Date of expiration now requires time to live in miliseconds.
  - Before:
    ```js
    await cajache.use(
      "id",
      () => functionToCache,
      {
        expire: (Date.now() / 1000) + 1, // 1 second
      }
    );
    ```
    
  - After:
    ```js
    await cajache.use(
      "id",
      () => functionToCache,
      {
        ttl: 1000, // 1 second
      }
    );
    ```
- `.expireWatcher` method has been removed, the watcher now starts automatically.



### Added
- New method `.deleteAll` has been added.
- New option `checkInterval` has been added. It defines the TTL watcher interval between cycles in miliseconds. Default value is `1000 * 60` (1 minute).
- New method `.new` has been added. It creates a new instance of `cajache`.
- New method `.setOptions` has been added. It sets the options for the current instance.



### Changed:

- The `condition` option behaviour been slighly changed:
  - **Before:** _Function that will receive as argument the fetchFnc response. If it returns true the response will be cached, if it returns an object it will override the response, otherwise it won't be cached and cajache.use will return null._
  - **After:** _On cache miss, this function will receive as argument the response of fnc. If it returns true the response will be cached, if it returns false it won't be cached._



### Fixed
- `cajache.delete` no longer deletes all cache entries with undefined id, now it throws an error.




tests/cajache.test.js

const cajache = require("../lib/cajache");



describe("use", () => {
	
	const generateRandomId = () => Math.random().toString(36).slice(2);
	
	const person = {
		money: 0,
		name: "Pepe",
		age: 123,
		metadata: {
			createdAt: "yesterday",
			updatedAt: "today",
		}
	};
	
	const addMoney = () => {
		person.money ++;
		return person.money;
	};
	
	const AddMoneyRetObj = () => {
		person.money ++;
		return person;
	};
	
	
	
	test("first, cached, fresh", async () => {
		
		let res = await cajache.use(
			"aaa",
			addMoney,
		);
		expect(res).toBe(1);
		
		
		res = await cajache.use(
			"aaa",
			addMoney,
		);
		expect(res).toBe(1);
		
		
		res = await cajache.use(
			"bbb",
			addMoney,
		);
		expect(res).toBe(2);
		
	});
	
	
	
	test("options.path", async () => {
		
		let res = await cajache.use(
			"path_1",
			AddMoneyRetObj,
			{
				path: "name",
			}
		);
		expect(res).toBe("Pepe");
		
	});
	
	test("options.path_deep", async () => {
		
		let res = await cajache.use(
			"path_2",
			AddMoneyRetObj,
			{
				path: "metadata.createdAt",
			}
		);
		expect(res).toBe("yesterday");
		
	});
	
	
	
	test("options.condition boolean", async () => {
		
		let res = await cajache.use(
			"condition_true_then_cache",
			AddMoneyRetObj,
			{
				condition: res => res.age === 123,
			}
		);
		expect(res.age).toBe(123);
		
		
		res = await cajache.use(
			"condition_false_then_nocache",
			AddMoneyRetObj,
			{
				condition: res => res.age === 0,
			}
		);
		expect(res.age).toBe(123);
		
	});
	
	
	
	test("delete", async () => {
		
		const response1 = await cajache.use(
			"first",
			generateRandomId,
		);
		
		cajache.delete("first");
	   
		const response2 = await cajache.use(
			"first",
			generateRandomId,
			{
				ttl: 1000 * 10,
			}
		);
		
		expect(response1).not.toBe(response2);
		
	});
	
	
	
	test("deleteAll", async () => {
		
		const response1 = await cajache.use(
			"first",
			generateRandomId,
		);
		
		cajache.deleteAll();
	   
		const response2 = await cajache.use(
			"first",
			generateRandomId,
			{
				ttl: 1000 * 10,
			}
		);
		
		expect(response1).not.toBe(response2);
		
	});
	
});

playground.js

const cajache = require("./lib/cajache");



(async() => {
    
    const generateRandomId = () => Math.random().toString(36).slice(2);
    
    
    console.time("fetch 1");
    const response1 = await cajache.use(
        "first",
        generateRandomId,
        {
            ttl: 1000 * 5,
        }
    );
    console.timeEnd("fetch 1");
    console.log(`    ${response1}`);
    
    
    
    console.time("fetch 2");
    const response2 = await cajache.use(
        "first",
        generateRandomId,
        {
            ttl: 100,
        }
    );
    console.timeEnd("fetch 2");
    console.log(`    ${response2}`);
    
    
    // Sleep 2s
    await new Promise(resolve => setTimeout(resolve, 1000));
    
    
    
    console.time("fetch 3");
    const response3 = await cajache.use(
        "first",
        generateRandomId,
        {
            ttl: 500,
        }
    );
    console.timeEnd("fetch 3");
    console.log(`    ${response3}`);
    
})();


playground2.js

const axios = require("axios");
const cajache = require("./lib/cajache");

const cajache2 = cajache.new();



(async() => {
    
    console.time("fetch 1");
    const response1 = await cajache2.use(
        "characters",
        () => axios.get("https://rickandmortyapi.com/api/character/14"),
    );
    console.timeEnd("fetch 1");
    console.log(`    ${response1.data?.name}`);

    
    console.time("fetch 2 (cached)");
    const response2 = await cajache2.use(
        "characters",
        () => axios.get("https://rickandmortyapi.com/api/character/14"),
    );
    console.timeEnd("fetch 2 (cached)");
    console.log(`    ${response2.data?.name}`);
    
    
    
    console.time("fetch 3 (cached)");
    const response3 = await cajache2.use(
        "characters",
        () => axios.get("https://rickandmortyapi.com/api/character/14"),
    );
    console.timeEnd("fetch 3 (cached)");
    console.log(`    ${response3.data?.name}`);
    
    
    
    // fetch 1: 248.659ms
    // fetch 2 (cached): 0.015ms
    // fetch 3 (cached): 0.008ms
    
})();


types/cajache.d.ts
export = defaultInstance;
declare const defaultInstance: Cajache;
/**
 * @callback ConditionFnc
 * Function that will receive as argument the `fetchFnc` response.
 * - If this function returns `true` the `fetchFnc` response will be cached.
 * - If this function returns `object` it will override the response.
 * - Otherwise `fetchFnc` response won't be cached and `cajache.use` will return null.
 * @param {*} fetchFncResponse Response of the `fetchFnc` function.
*/
/**
 * @typedef Config
 * @property {number} ttl Time to live in milseconds.
 * @property {number} checkInterval Default `1000 * 60` (1 min). Interval in milseconds to check for expired entries. **Only used on instance config.**
 * @property {string} path Dot path to the property that will be saved. Example: `"user.data"`.
 * @property {ConditionFnc} condition Function that will return boolean determining if the response will be cached or not.
*/
declare class Cajache {
    /**
     * @param {Config} config
    */
    constructor(config?: Config);
    id: string;
    config: {};
    cache: {};
    expiringEntries: Set<any>;
    /**
     * Retrieves a entry.
     * @param {string | Array<string>} id Unique ID of the cache entry.
     * @returns {* | null}
    */
    get(id: string | Array<string>): any | null;
    /**
     * Sets the data for a cache entry.
     * @param {string | Array<string>} id Unique ID of the box.
     * @param {*} data Data you want to set.
     * @param {Config} options
    */
    set(id: string | Array<string>, data: any, options?: Config): void;
    /**
     * Deletes a cache entry.
     * @param {string | Array<string>} id Unique ID of the cache entry.
     * @returns {*} Returns the deleted cache entry.
    */
    delete(id: string | Array<string>): any;
    /**
     * Deletes all cache entries.
    */
    deleteAll(): void;
    /**
     * @param {string | Array<string>} id Unique ID of the cache entry.
     * @param {function} fnc Function that will be executed (and saved) on cache miss, otherwise it will just return the cached data.
     * @param {Config} options
     *
     * @returns {*} Returns the cached result if found, otherwise returns the returned data of `fnc`.
    */
    use(id: string | Array<string>, fnc: Function, options?: Config): any;
    /**
     * Creates a new instance of cache, with his own config and separated cache.
     * @param {Config} config
     * @returns {Cajache}
    */
    new(config: Config): Cajache;
    /**
     * Sets config values
     * @param {"ttl" | "condition" | "path" | "checkInterval"} key
    */
    setConfig(key: "ttl" | "condition" | "path" | "checkInterval", value: any): void;
}
type Config = {
    /**
     * Time to live in milseconds.
     */
    ttl: number;
    /**
     * Default `1000 * 60` (1 min). Interval in milseconds to check for expired entries. **Only used on instance config.**
     */
    checkInterval: number;
    /**
     * Dot path to the property that will be saved. Example: `"user.data"`.
     */
    path: string;
    /**
     * Function that will return boolean determining if the response will be cached or not.
     */
    condition: ConditionFnc;
};
/**
 * Function that will receive as argument the `fetchFnc` response.
 * - If this function returns `true` the `fetchFnc` response will be cached.
 * - If this function returns `object` it will override the response.
 * - Otherwise `fetchFnc` response won't be cached and `cajache.use` will return null.
 */
type ConditionFnc = (fetchFncResponse: any) => any;


lib/cajache.js

/**
 * Obtiene la referencia de un objeto anidado dentro de otro objeto (a cualquier profundidad) a partir de una ruta `["key1", "key2", "key3"...]`.
 * @param {*} obj 
 * @param {Array<string> | string} arrKeys Ruta hasta el objeto. Por ejemplo `["moto", "rueda2", "tipo"]`.
 * @returns {Array<*,string>} [ref, lastKey] para poder hacer `ref[lastKey] = "cosas"`.
 * 
 * @example
 * 
 * let moto = {
 * 	rueda1: {
 * 		tipo: "AD-56",
 * 		estado: "ok"
 * 	},
 * 	rueda2: {
 * 		tipo: "AT-77",
 * 		estado: "ok"
 * 	},
 * };
 * 
 * let [ref, lastKey] = getRef(moto, ["rueda2", "tipo"]);
 * console.log( ref[lastKey] ); // "AT-77"
 * 
*/
function getRef(obj, arrKeys) {
	
	if (typeof arrKeys === "string") return [obj, arrKeys];
	
	
	let ref = obj;
	let lastKey = null;
	
	
	// Si me pasan array
	const lastIdx = arrKeys.length - 1;
	
	arrKeys.forEach( (_key, _idx) => {
		if (_idx === lastIdx) return lastKey = _key;
		
		if (!ref[_key]) ref[_key] = {};
		ref = ref[_key];
	});
	
	
	return [ref, lastKey];
	
};



// -------------------------------------------------------------------------------------------------------------------------------



const startWatcher = (instance) => {
	
	setTimeout( () => {
		
		for (let _cacheId of instance.expiringEntries) {
			
			const [ref, lastKey] = getRef(instance.cache, _cacheId);
			const cacheEntry = ref[lastKey];
			
			if (cacheEntry && cacheEntry.expireAt && cacheEntry.expireAt < Date.now()) {
				delete ref[lastKey]; // borro el cache
				instance.expiringEntries.delete(_cacheId); // y lo borro de la lista
			};
			
		};
		
		const checkInterval = instance?.config?.checkInterval ?? 0;
		if (checkInterval <= 0) return; // interval is 0, stop
		startWatcher(instance); // start again
		
	}, instance?.config?.checkInterval ?? 0);
	
};



const defaultCajacheConfig = {
	ttl: 0, // 0 = no ttl
	checkInterval: 1000 * 60, // 1 min
	path: null,
	condition: null,
}


/**
 * @callback ConditionFnc
 * Function that will receive as argument the `fetchFnc` response.
 * - If this function returns `true` the `fetchFnc` response will be cached.
 * - If this function returns `object` it will override the response.
 * - Otherwise `fetchFnc` response won't be cached and `cajache.use` will return null.
 * @param {*} fetchFncResponse Response of the `fetchFnc` function.
*/

/**
 * @typedef Config
 * @property {number} ttl Time to live in milseconds.
 * @property {number} checkInterval Default `1000 * 60` (1 min). Interval in milseconds to check for expired entries. **Only used on instance config.**
 * @property {string} path Dot path to the property that will be saved. Example: `"user.data"`.
 * @property {ConditionFnc} condition Function that will return boolean determining if the response will be cached or not.
*/

class Cajache {
	id = "cajache";
	config = {};
	cache = {};
	expiringEntries = new Set();
	
	/**
	 * Retrieves a entry.
	 * @param {string | Array<string>} id Unique ID of the cache entry.
	 * @returns {* | null}
	*/
	get(id) {
		try {
			let [ref, lastKey] = getRef(this.cache, id);
			
			
			// No lo he encontrado
			if (!ref[lastKey]) return null;
			
			
			// Si está caducado, borro la entrada y devuelvo null, como si no lo hubiese encontrado
			if (ref[lastKey].ttl < ( Date.now() / 1000)) {
				delete ref[lastKey];
				return null;
			};
			
			// Si llego hasta aquí, lo he encontrado y devuelvo su propiedad data
			return ref[lastKey].data;
			
		} catch(err) {
			console.error(err);
			return null;
		};
	};
	
	/**
	 * Sets the data for a cache entry.
	 * @param {string | Array<string>} id Unique ID of the box.
	 * @param {*} data Data you want to set.
	 * @param {Config} options
	*/	
	set(id, data, options = {}) {
		try {
			let [ref, lastKey] = getRef(this.cache, id);
			
			if (!ref[lastKey]) ref[lastKey] = {}; // si no existe la caja, la creo
			ref[lastKey].data = data; // meto data
			
			
			if (options.ttl) {
				ref[lastKey].expireAt = Date.now() + options.ttl;
				
				const stringifiedId = typeof id === "string" ? id : id.join(".");
				this.expiringEntries.add(stringifiedId);
			};
			
		} catch (err) {
			console.error(err);
		};
	};
	
	/**
	 * Deletes a cache entry.
	 * @param {string | Array<string>} id Unique ID of the cache entry.
	 * @returns {*} Returns the deleted cache entry.
	*/
	delete (id) {
		try {
			if (!id) throw new Error(`delete should have a valid id but ${id} was provided`); 
			
			let [ref, lastKey] = getRef(this.cache, id);
			
			let deleted = ref[lastKey];
			delete ref[lastKey];
			
			return deleted;
		} catch(err) {
			console.error(err);
			return null;
		};
	};
	
	/**
	 * Deletes all cache entries.
   	*/	
	deleteAll() {
		this.cache = {};
	};
	
	/**
	 * @param {string | Array<string>} id Unique ID of the cache entry.
	 * @param {function} fnc Function that will be executed (and saved) on cache miss, otherwise it will just return the cached data.
	 * @param {Config} options
	 * 
	 * @returns {*} Returns the cached result if found, otherwise returns the returned data of `fnc`.
	*/
	async use (id, fnc, options = {}) {
		
		try {
			// Find cache
			const cache = this.get(id);
			
			// Cache hit
			if (cache) {
				this.set(id, cache, options); // set new options
				return cache;
			};
			
			// Cache miss, execute fnc
			let freshData = await fnc();
			
			// Get config
			const {condition, path} = options ?? this.config;
			
			// Check condition
			if (condition) {
				if (typeof condition !== "function") throw TypeError("config.condition must be a function");
				
				const conditionResult = condition(freshData); // try to execute condition
				if (!!conditionResult !== true) return freshData; // condition check failed, return freshData and don't cache
				
			};
			
			// Path is required
			if (path) {
				const [ref, lastKey] = getRef(freshData, path.split("."));
				freshData = ref[lastKey];
			};
			
			
			this.set(id, freshData, this.options);
			return freshData;
			
		} catch (err) {
			console.error(err);
			return null;
		}
		
		
	};
	
	/**
	 * Creates a new instance of cache, with his own config and separated cache.
	 * @param {Config} config
	 * @returns {Cajache}
	*/
	new(config) {
		return new Cajache(config);
	};
	
	/**
	 * Sets config values
	 * @param {"ttl" | "condition" | "path" | "checkInterval"} key
	*/
	setConfig(key, value) {
		
		if (key === "ttl") this.config.ttl = value;
		if (key === "path") this.config.path = value;
		if (key === "condition") this.config.condition = value;
		if (key === "checkInterval") {
			this.config.checkInterval = ms;
			if (ms > 0) startWatcher(this);
		};
	};
	
	/**
	 * @param {Config} config
	*/
	constructor(config = defaultCajacheConfig) {
		this.id = Math.random().toString(36).slice(2);
		this.config = config;
		
		startWatcher(this);
		
		return this;
	};
	
};


const defaultInstance = new Cajache();


module.exports = defaultInstance;
module.exports.default = defaultInstance;

